File: templates/python/provider.py
#!/usr/bin/python

import json
import os
import requests
import time

from posixpath import join as urljoin

PROVIDER_NAME = "<PROVIDER NAME>"
ENDPOINT_REGISTER = "providers"
ENDPOINT_PUSH = "notifications"

headers = {'content-type': 'application/json'}
visitedStories = {}

def register(shruti_server):
    _data = { "name" : PROVIDER_NAME,
              "display_name" : "<DISPLAY NAME>",
              "description": "<DESCRIPTION>",
              "web_url": "<URL>",
              "icon_url": "<URL>",
              "voice": "<VoiceName>"
        }
    _url = urljoin(shruti_server, ENDPOINT_REGISTER, PROVIDER_NAME)
    print _url
    r = requests.post(_url, data=json.dumps(_data), headers=headers)
    if r.text == 'pq: duplicate key value violates unique constraint "providers_name_key"':
        return 201, "already registered"
    return r.status_code, r.text

def doWork(shruti_server):

    _url = urljoin(shruti_server, ENDPOINT_PUSH)
    print _url

    while True:
        _resp = requests.get("<IMPLEMENT THIS>")

        #submit story to /notifications
        _data = { "title" : "<TITLE TEXT>",
                  "url": "<URL>",
                  "key": PROVIDER_NAME + "<KEY>",
                  "priority": 20,
                  "action": 10,
                  "provider_name": PROVIDER_NAME
                  }
        r = requests.post(_url, data=json.dumps(_data), headers=headers)
        visitedStories["<KEY>"] = True

        msg = "Error submitting story:" + r.text if r.status_code == 500 else "submitted"
        print msg

        time.sleep(15 * 60) # ~15minutes
        print "time to check updates again"

def main():
    shruti_server = None
    shruti_server = os.environ['SHRUTI_SERVER']
    if shruti_server is None:
        print "server url not set"
        return
    status, resp = register(shruti_server)
    print status, resp
    if status == 500:
        print "Error registering:", resp
        return
    doWork(shruti_server)

if __name__ == "__main__":
    main()

File: templates/go/provider.go
package main

import (
	"log"
	"os"
	"time"

	//"github.com/SlyMarbo/rss"
	shruti "github.com/omie/shruti-go"
)

const (
	PROVIDER_NAME = "<PROVIDER_NAME>"
)

var (
	sClient shruti.Client
	visited map[string]bool
)

func Register() (err error) {

	p := shruti.Provider{Name: PROVIDER_NAME,
		DisplayName: "<DisplayName>",
		Description: "<Discription>",
		WebURL:      "<WebURL>",
		IconURL:     "<IconURL>",
		Voice:       "<Voice>",
	}

	err = sClient.RegisterProvider(p)

	return

}

func doWork() (err error) {

	for {
		for _, item := range items[:10] {
			if _, ok := visited[item.ID]; ok {
				continue
			}
			n := shruti.Notification{
				Title:        item.Title,
				Url:          item.Link,
				Key:          PROVIDER_NAME + item.ID,
				Priority:     shruti.PRIO_MED,
				Action:       shruti.ACT_PUSH,
				ProviderName: PROVIDER_NAME,
			}
			err = sClient.PushNotification(n)
			msg := "submitted"
			if err != nil {
				msg = err.Error()
			}
			visited[item.ID] = true
			log.Println(msg)
		}
		log.Println("Sleeping")
		time.Sleep(15 * time.Minute)
		log.Println("woke up")

	}
}

func main() {

	log.Println("Starting Scraper")

	shrutiServer := os.Getenv("SHRUTI_SERVER")
	if shrutiServer == "" {
		log.Println("SHRUTI_SERVER not set")
		return
	}

	sClient = shruti.Client{"http://", shrutiServer}

	err := Register()
	if err != nil {
		log.Println(err)
		return
	}

	visited = make(map[string]bool)

	err = doWork()
	if err != nil {
		log.Println(err)
	}
}
